### PR TITLE
[SPARK-32119][CORE][3.0] ExecutorPlugin doesn't work with Standalone Cluster and Kubernetes with --jars

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This is a backport PR for branch-3.0.

This PR changes Executor to load jars and files added by --jars and --files on Executor initialization.
To avoid downloading those jars/files twice, they are assosiated with `startTime` as their uploaded timestamp.

### Why are the changes needed?

ExecutorPlugin can't work with Standalone Cluster and Kubernetes
when a jar which contains plugins and files used by the plugins are added by --jars and --files option with spark-submit.

This is because jars and files added by --jars and --files are not loaded on Executor initialization.
I confirmed it works with YARN because jars/files are distributed as distributed cache.

### Does this PR introduce _any_ user-facing change?

Yes. jars/files added by --jars and --files are downloaded on each executor on initialization.

### How was this patch tested?

Added a new testcase.

### CODE REVIEW

**Comment by sarutak** [comment]
cc: @tgravescs @mridulm @LucaCanali

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
cc @zhengruifeng since you're the release manager of 3.0.1

**Comment by zhengruifeng** [comment]
@HyukjinKwon Thanks for letting me know this. It seems that this issue may also exist in 2.4? I think you can vote -1 in this active vote.

**Comment by HyukjinKwon** [comment]
Assuming from the JIRA, seems like it's not a blocker. I just cc'ed you just as a FYI :-). but @sarutak can you confirm if this blocks the release?

**Comment by sarutak** [comment]
I don't think, this is a blocker. We can't deploy plugins and its required files with --jars/--files but as @tgravescs mentioned [here](https://issues.apache.org/jira/browse/SPARK-32119?focusedCommentId=17148746&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17148746), we have a workaround by installing plugins on all worker nodes beforehand and specifying them with extraClassPath.

**Comment by dongjoon-hyun** [comment]
Ya. I agree that this is not a blocker for 3.0.1 release.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Review by tgravescs — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
AmbLap Server seems to have a broken `Maven` local repository again.
```
========================================================================
Running build tests
========================================================================
exec: curl -s -L https://downloads.lightbend.com/zinc/0.3.15/zinc-0.3.15.tgz
exec: curl -s -L https://downloads.lightbend.com/scala/2.12.10/scala-2.12.10.tgz
Using `mvn` from path: /home/jenkins/tools/hudson.tasks.Maven_MavenInstallation/Maven_3.6.3/bin/mvn
Using `mvn` from path: /home/jenkins/tools/hudson.tasks.Maven_MavenInstallation/Maven_3.6.3/bin/mvn
Performing Maven install for hadoop-2.7-hive-1.2
Using `mvn` from path: /home/jenkins/tools/hudson.tasks.Maven_MavenInstallation/Maven_3.6.3/bin/mvn
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-install-plugin:3.0.0-M1:install (default-cli) on project spark-launcher_2.12: ArtifactInstallerException: Failed to install metadata org.apache.spark:spark-launcher_2.12/maven-metadata.xml: Could not parse metadata /home/jenkins/.m2/repository/org/apache/spark/spark-launcher_2.12/maven-metadata-local.xml: in epilog non whitespace content is not allowed but got > (position: END_TAG seen ...</metadata>\n>... @13:2) -> [Help 1]
[ERROR] 
```

**Comment by dongjoon-hyun** [comment]
This PR doesn't pass K8s IT since September.
- https://github.com/apache/spark/pull/29621#issuecomment-685170312
- https://github.com/apache/spark/pull/29621#issuecomment-685235211
- https://github.com/apache/spark/pull/29621#issuecomment-718138054
- https://github.com/apache/spark/pull/29621#issuecomment-718196120
- https://github.com/apache/spark/pull/29621#issuecomment-718196579

The failures are on the following test cases. Recently, Jenkins K8s IT seems to be unstable.
- Run SparkPi with no resources (Failed always, but known to be flaky)
- Run SparkPi with a very long application name. (failed once in the above)
- Run SparkR on simple dataframe.R example (Failed always, but known to be flaky

So, I verified manually on my laptop.
```
ubernetesSuite:
- Run SparkPi with no resources
- Run SparkPi with a very long application name.
- Use SparkLauncher.NO_RESOURCE
- Run SparkPi with a master URL without a scheme.
- Run SparkPi with an argument.
- Run SparkPi with custom labels, annotations, and environment variables.
- All pods have the same service account by default
- Run extraJVMOptions check on driver
- Run SparkRemoteFileTest using a remote data file
- Run SparkPi with env and mount secrets.
- Run PySpark on simple pi.py example
- Run PySpark with Python2 to test a pyfiles example
- Run PySpark with Python3 to test a pyfiles example
- Run PySpark with memory customization
- Run in client mode.
- Start pod creation from template
- PVs with local storage
- Launcher client dependencies
- Run SparkR on simple dataframe.R example
Run completed in 10 minutes, 17 seconds.
```

For the K8s IT failure, I've been taking a look independently.
- https://github.com/apache/spark/pull/30130

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to branch-3.0 for Apache Spark 3.0.2.

**Comment by sarutak** [comment]
@dongjoon-hyun @tgravescs Thank you for taking look at this PR again!

**Pull request closed without merging**
